---
title: "04_Plotting_TPM"
author: "JR"
date: "2024-08-04"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
install.packages("reshape")
library(reshape)
```

# Goal: use TPM values to make plots of how genes change in abundance due to dox and time
So far we have been learning to plot using the results of DESEQ2 such as pvalue, padj 
Now we are going to take another step in plotting the abundance of each gene (TPM).
This would be the type of plot you do after finding sig genes in DESQ2 to see the trend of 
abundance and time. Let's use our figure skills and learn some more R and ggplot2 to measure gene expression levels.

# Important: Counts are a much more accurate representation of how much an RNA was observed in a sample
TPM compresses all those thousands of reads on each gene into one value (Transcripts per million).
TPM is just simply saying what percentage of a given sequencing sample a given gene is.
We also saw that almost all our data is at value TPM < 20 and std is very high -
So take all of this with a grain of salt but is good to measure relative changes in gene across samples!

# Let's start making some TPM plots !

# Load data
```{r loading our RNAseq analysis data TPM & COUNTS}

# load TPM RESULTS: 
load("/scratch/Shares/rinnclass/MASTER_CLASS/lessons/06_Differential_expression_analyses/results/TPM_results/TPM_results.Rdata")

# laod DESEQ2 results
load("/scratch/Shares/rinnclass/MASTER_CLASS/lessons/06_Differential_expression_analyses/results/counts_results_DESEQ/DESEQ_results.rdata")

```

# First let's revisit the distribution of values for TPM 
```{r hist TPM values}
# Previously we plotted TPM distribution from our cols labeled:
# "WT_0_1" etc - note that is not a good idea to name a col by a number :)
hist(average_and_stddev_values$96_avg)
# Oh no this doesn't work because of an "integer" value being indexed
# This can be fixed easily by :
hist(average_and_stddev_values$'96_avg', xlim = c(0,1000), ylim = c(0,1000), breaks = 1000)
# We can quickly see again that most values are very low and a long right tail on distribution of TPM values!
```

# Now let's do this with Ggplot and learn to melt data :) Or pivot longer
Our goal here is to make a distribution of TPM values in each time point - using facet-wrap
```{r facet all timepoint TPM distributions}

# First let's make set of time values for faceting:
time_values <- c("0_avg", "12_avg", "24_avg", "48_avg", "96_avg")

# This can also be done with grep (smarter):
time_points <- names(average_and_stddev_values)[grep("avg", names(average_and_stddev_values))]
# moving rows into df so we can track gene_id
average_and_stddev_values_2 <- rownames_to_column(average_and_stddev_values, var = "gene_id")

# Melt the data frame to long format
melt_tpm_df <- melt(average_and_stddev_values_2, measure.vars = time_points, value_name = "gene_id")
# Cool we just made it a really long df (compare original and melt) -it's 5x longer

# Now to plot:
ggplot(melt_tpm_df, aes(x = value)) +
  geom_histogram(bins = 50) +
  facet_wrap(~ variable, scales = "free_x")+
  xlim(0,1000) +
  ylim(0,1000) +
  theme("paperwhite")

# Just for fun we can also get some of this in summary
summary(average_and_stddev_values)
```
# RESULT: average TPM distributions are similar in all samples


# Let's grab the counts just for the filtered res_df genes.
# Also let's make sure we have a list of unique genes!
```{r TPM for DESEQ2 sig genes}

sig_genes <- as.data.frame(unique(filtered_res_df$gene_id))
names(sig_genes) <- "gene_id"

# one way to subset the data frame of avrg / sd TPM values
sig_gene_tpm <- inner_join(average_and_stddev_values_2, sig_genes, by = "gene_id")

# Same can be done by indexing
sig_gene_tpm_2 <- average_and_stddev_values_2[average_and_stddev_values_2$gene_id %in%
                                                sig_genes$gene_id, ]

# let's now merge in the gene names into this handy dataframe
sig_gene_tpm <- merge(sig_gene_tpm, g2s)

# Now we should organize so we can plot specific genes (not mix character and integer)
sig_gene_tpm <- sig_gene_tpm %>%
  select(-gene_id) 

# cool now we have the tpm and sd values to plot !

```


# Now let's plot one gene TPM and sd across time, let's look in df for a good one (filtered_res_df)
```{r TPM plot of gene of interest}

# Picking a favorite gene
gene_up <- as.data.frame(c("Gm9045"))
names(gene_up) <- "gene_name"

# indexing to just get values for this gene
Gm9045_tpm <- sig_gene_tpm[sig_gene_tpm$gene_name %in% gene_up$gene_name, ]

# now moving the character gene_name to rownames to won't interfere with numeric values
rownames(Gm9045_tpm) <- Gm9045_tpm$gene_name

# removing text column
Gm9045_tpm <- Gm9045_tpm %>%
  select(-gene_name) 

# Ok this is typical data wrangling now let's plot !

```

# Line plot for TPM and SD across time points
```{r plotting line plot}

# We now need to futher prepare our data :)
# Let's get a variable for col names to plot
avg_cols <- grep("_avg", names(Gm9045_tpm), value = TRUE)
sd_cols <- grep("_sd", names(Gm9045_tpm), value = TRUE)

# This is another way to "melt" or pivot longer
plot_df <- data.frame(
  time = as.numeric(gsub("_avg", "", avg_cols)),
  avg = unlist(Gm9045_tpm[avg_cols]),
  sd = unlist(Gm9045_tpm[sd_cols])
)

# NOW THE PLOT !
ggplot(plot_df, aes(x = time, y = avg)) +
  geom_line() +
  geom_errorbar(aes(ymin = avg - sd, ymax = avg + sd), width = 0.2) +
  theme("paperwhite") +
  labs(x = "Time (hours)", y = "TPM Value")
# +   theme_classic()

```



#TODO start here

```{r}
#TODO start here 

genes_of_interest <- data.frame(gene = c("ENSMUSG00000094125.1", "ENSMUSG00000079247.2", "ENSMUSG00000112027.1", "ENSMUSG00000066632.3"))

df_plot_genes <- TPM_filtered[rownames(TPM_filtered) %in% genes_of_interest$gene, ]
# moving rownames in
df_plot_genes <- rownames_to_column(df_plot_genes, var = "gene_id")


# TODO start here






df_plot_genes <- df_plot_genes %>%
  pivot_longer(cols = everything(),
               names_to = c("condition", "time", "replicate"),
               names_sep = "_")


df_plot_genes$condition <- gsub("WT_", "", df_plot_genes$condition)
df_plot_genes$condition <- as.numeric(df_plot_genes$condition)
df_plot_genes$replicate <- as.numeric(df_plot_genes$replicate)


library(ggplot2)
ggplot(df_plot_genes, aes(x = condition, y = value, color = factor(replicate))) +
  geom_line(aes(group = interaction(gene, replicate))) +
  geom_point() +
  labs(x = "Time (hours)", y = "TPM Value") +
  theme_classic()

```







