---
title: "02_peaks_and_promoters"
author: "JR"
date: "2024-11-21"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# GOAL : #################

```{r setting up information to load peak files}

# sourcing import peaks function 
source("../../util/00_our_first_function_lesson.R")

# establishing peak path to the dir with MACS2 output peak files from NF_CORE ATACseq pipeline
peak_path <- "/scratch/Shares/rinnclass/MASTER_CLASS/lessons/08_ATACseq_pipeline/00_run_pipeline/00_pipeline_run/pipeline_run/bwa/merged_library/macs2/broad_peak"

# creating a file list also needed for import_peaks function
fl <- list.files(peak_path, full.names = TRUE, pattern = ".broadPeak")
```

# Run import_peaks function to get a list of ATAC peak files in GRanges format

```{r run import_peaks}

my_peaks <- import_peaks(consensus_file_path = peak_path)

# Nice so with a list of peak files and path to the folder we can load as many peak files as we need !
num_peaks <- sapply(my_peaks, length) %>% as.data.frame
view(num_peaks)

# printing to compare later to filtered to cannonical chromosomes  
print(num_peaks)

```

# Now let's find a set of peaks that are common to all the samples !
We previously looked at the overlap between WT_0 and KO_O now let's find the peaks that are in all files.
Let's go through each step and then 


# Step one Genomic ranges reduce function across the list of GRanges for each ATACseq peak file: 
The GenomicRanges::reduce function in the GenomicRanges package merges overlapping or adjacent genomic ranges into a single range, depending on a specified rule. 
# By default, it takes the outermost boundaries of all overlapping or adjacent ranges, creating the smallest set of non-overlapping ranges that span all the input ranges.

# Let's put it to use and find peaks in all time points !
```{r all time point consensus peaks}
# making a list of peak files 
fl <- list.files(peak_path, 
                 full.names=TRUE, pattern = ".broadPeak")

# Extracting sample name from file name
sample_name <- sapply(fl, function(y){
 y <-  str_extract(y, "(KO|WT)_control_\\d+")
  
})

# Now we have the needed inputs to make a function !
# Let's make a fucntion to create peaks in all input files.
# Also to count how many peaks are in each file (number of overlaps)
```


```{r}

find_common_peaks <- function(peak_list) {
  # Validate input
  if (!is.list(peak_list) || !all(sapply(peak_list, inherits, "GRanges"))) {
    stop("Input must be a list of GRanges objects.")
  }
  
  # Start with the first GRanges object
  common_peaks <- peak_list[[1]]
  
  # Iteratively find overlaps across all GRanges objects
  for (i in 2:length(peak_list)) {
    current_gr <- peak_list[[i]]
    
    # Find overlaps
    overlaps <- findOverlaps(common_peaks, current_gr)
    
    # Subset to overlapping regions
    common_peaks <- subsetByOverlaps(common_peaks, current_gr)
  }
  
  # Assign custom names to the common peaks
  mcols(common_peaks)$name <- paste0("common_peak_", seq_along(common_peaks))
  
  return(common_peaks)
}

# Let's give it a test run !

common_peaks <- find_common_peaks(my_peaks)

```

# Take a moment and make this function in your .Rscrip in util

Now back to our peaks. Let's take a look at what our common peaks looks like
relative to an individual peak file (or two :).

# (1) We need to make our common peaks object into a .BED file to view in IGV.
Let's make a handy script to make peak objects in environment into viewable files !

```{r}

output_bed <- "common_peaks.bed"

# Export the GRanges object to a BED file
rtracklayer::export(common_peaks, con = output_bed, format = "BED")

cat("Exported common peaks to:", output_bed, "\n")


```


# (2) Load in the common peaks and individual ATAC bigwig files to IGV and investigate:
(a) Do the peaks look like they are common in all samples?
(b) search around for good peaks and note some examples (the coordinates on IGV - put in .RMD)
(c) What is weird about Firre? 
(d) how many peaks are in your common_peaks file versus the one produced by NF_CORE pipeline?
  - Hint <output_dir> /bwa/merged_library/macs2/broadPeak/consensus/consensus_peaks.mLb.clN.boolean.intersect.plot.pdf

# (3) Now load in the consensus peak file that has any peak in any sample and compare in IGV
(a) Find a region where there is a common peak next to a peak that is not common but called by pipeline
(b) Look at some common peaks and see how often a consensus_peak is present - would expect almost all right?




























