---
title: "04_run_Deseq2"
author: "JR"
date: "2024-04-16"
output: html_document
---
```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)
library(IRanges)
library(dplyr)
library(tidyr)
library(readr)
library(ggplot2)
library(purrr)
library(magrittr)
library(pheatmap)
library(textshape)
library(Rcpp)

#install.packages("textshape")
# install.packages("Rcpp")


#TODO ADD VIDEO
library(DESeq2)

```

```{R} 


# Load files .Rdata

##############################
##############################DESEQ2
    
dds <- DESeqDataSetFromMatrix(countData = counts_filtered,
                              colData = deseq_samples,
                              design = ~ time_point + replicate)

dds <- DESeq(dds)

#TODO result names
results <- results(dds)

significant_results <- results[which(results$padj < 0.05), ]






#####################

dds_2 <- DESeqDataSetFromMatrix(countData = counts_filtered,
                              colData = deseq_samples,
                              design = ~ time_point)

dds_2 <- DESeq(dds_2)
results_2 <- results(dds_2)

significant_results_2 <- results[which(results_2$padj < 0.05), ]

#######################################

normalized_counts <- counts(dds, normalized = TRUE)
gene_of_interest <- "ENSMUSG00000021270.13"
gene_expression <- as.data.frame(normalized_counts[gene_of_interest, ])
gene_expression$sample <- rownames_to_column(gene_expression)




gene_expression_df <- data.frame(TimePoint = rownames(gene_expression),
                                 Expression = as.numeric(gene_expression))


ggplot(gene_expression, aes(x = TimePoint, y = Expression)) +
  geom_line() +
  labs(x = "Time Point", y = "Expression Level", title = paste("Expression Profile of", gene_of_interest)) +
  theme_minimal()



significant_results <- results[which(results$padj < 0.05), ]

gene_of_interest <- rownames(significant_results)
gene_of_interest <- "ENSMUSG00000021270.13"

gene_expression <- normalized_counts[gene_of_interest, ]
gene_expression_df <- data.frame(TimePoint = colnames(gene_expression),
                                 Expression = as.numeric(gene_expression))








gene_expression_df <- data.frame(TimePoint = colnames(gene_expression),
                                 Expression = as.numeric(gene_expression))


# results <- results(dds)

expression_data <- assay(dds)
expression_data_df <- as.data.frame(t(expression_data))
expression_data_df$sample_id <- rownames(expression_data_df)
expression_data_long <- gather(expression_data_df, TimePoint, Expression, -sample_id)


expression_data_with_time_significant <- subset(expression_data_long, TimePoint %in% rownames(significant_genes))


ggplot(expression_data_with_time_significant, aes(x = , y = Expression)) +
  geom_line() +
  facet_wrap(~TimePoint, scales = "free_y")  # Facet by gene for individual plots




expression_data_with_time <- merge(expression_data_long, deseq_samples, by.x = "Sample", by.y = "sample_id", all.x = TRUE)






expression_data_with_time <- merge(expression_data_df, deseq_samples, by.x = "Sample", by.y = "sample_id", all.x = TRUE)




expression_data_df$Sample <- rownames(expression_data_df)

expression_data_with_time <- merge(expression_data_df, deseq_samples, by.x = "Sample", by.y = "sample_id", all.x = TRUE)

expression_data_long <- expression_data_with_time %>%
  pivot_longer(cols = starts_with("WT_"), names_to = "TimePoint", values_to = "Expression")




expression_data_long <- expression_data_with_time %>%
  pivot_longer(


expression_data$Sample <- rownames(expression_data)
expression_data_df <- as.data.frame(t(expression_data))

expression_data_with_time <- merge(expression_data_df, deseq_samples, by.x = "Sample", by.y = "sample_id", all.x = TRUE)


expression_data_long <- expression_data_with_time %>%
  pivot_longer(cols = starts_with("X"), names_to = "TimePoint", values_to = "Expression")




expression_data_long <- expression_data_with_time %>%
  pivot_longer(cols = starts_with("WT"), names_to = "TimePoint", values_to = "Expression")


significant_genes <- subset(results, padj < 0.05)

expression_data <- t(expression_data)
expression_data$Gene <- rownames(expression_data)


# GOES WRONG HERE
expression_data <- as.data.frame(expression_data)

expression_data_long <- expression_data %>%
  gather(TimePoint, Expression, -Gene)











#expression_data$Gene <- rownames(expression_data)








expression_data <- as.data.frame(expression_data)
expression_data$Gene <- rownames(expression_data)


expression_data_long <- expression_data %>%
  pivot_longer(cols names_to = "TimePoint", values_to = "Expression")





significant_genes_plot <- subset(results, padj < 0.05)

# Extract the gene IDs of significant genes
significant_gene_ids <- rownames(significant_genes_plot)


expression_data_significant <- expression[expression_data %in% significant_gene_ids, ]



# expression_data <- t(expression_data)


expression_data$Gene <- rownames(expression_data)

expression_data <- as.data.frame(expression_data)


expression_data_long <- expression_data %>%
  pivot_longer(cols = -Gene, names_to = "TimePoint", values_to = "Expression")


significant_genes_plot <- subset(results, padj < 0.05)

# Extract the gene IDs of significant genes
significant_gene_ids <- rownames(significant_genes_plot)

# Filter expression data to include only significant genes
expression_data_significant <- expression_data_long[expression_data_long$Gene %in% significant_gene_ids, ]

# Plot expression profiles of significant genes using ggplot2
ggplot(expression_data_significant, aes(x = TimePoint, y = Expression, color = Gene)) +
  geom_line() +
  labs(x = "Time Point", y = "Expression Level", title = "Expression Profiles of Significant Genes") +
  theme_minimal() +
  facet_wrap(~Gene, scales = "free_y")  # Facet by gene for individual plots













rownames(significant_results$rownames)

gene_of_interest <- "ENSMUSG00000021270.13"

expression_data_gene <- expression_data_long[expression_data_long$Gene == gene_of_interest, ]

ggplot(expression_data_gene, aes(x = TimePoint, y = Expression)) +
  geom_line() +
  labs(x = "Time Point", y = "Expression Level", title = paste("Expression Profile of", gene_of_interest)) +
  theme_minimal()










ggplot(expression_data_long, aes(x = TimePoint, y = Expression, color = Gene)) +
  geom_line() +
  labs(x = "Time Point", y = "Expression Level", color = "Gene") +
  theme_minimal() +
  theme(legend.position = "none") +
  facet_wrap(~Gene, scales = "free_y")  # Facet by gene for individual plots







```

Cool, now our counts sorted, but we need to connect to a sample sheet
DESeq requires a sample sheet to know what samples and replicates are
Let's make that now

```{r smaple sheet}

# First let's load a sample sheet containing samples/replicates
# We make a condition column so DEseq will treat them as in same "group"

samplesheet <- read_csv("/scratch/Shares/rinnclass/MASTER_CLASS/lessons/04_RNAseq_Dox/01_Mouse_dox_wt/data/samplesheet.csv")

# First let's put the counts columns in the same order as the samplesheet
# If they are not then we would essentially be randomizing our data later!!
# counts <- counts_filtered[,samplesheet$sample_id]

# Now we know they're in the same order, and we can change the names
# It is nicer to have more descriptive names.
colnames(counts_filtered) <- samplesheet$sample

# This is a handy check point to make sure the labels in 
# sample sheet and counts are similar
all(colnames(counts_filtered) == samplesheet$sample)

# Now our samplesheet and counts tables are organized the same

# Double check
colnames(counts_filtered) <- samplesheet$sample

# Nice!






```

# goal: to get rlog normalized counts 
Let's officially run DeSeq2 these are the steps we will take

(1) create an "design" for Deseq as input to create a 
"dds" (DESeqDataSet) :: "DESeqDataSetFromMatrix".
This comes from sample sheet.



# --- sample sheet needs to be factored! ----

```{R factor sample sheet}
#TODO Make a column for group to factor for DESEQ2
samplesheet$group <- samplesheet$sample 

samplesheet$group <- strsplit(samplesheet$group, split = "_")
?stringsp

#TODO Make a column for group to factor for DESEQ2



# A FACTOR LEVEL is critical for DEseq2 to know which samples is which
# We will take the condition from sample sheet to compare for differential expression
# In this case it doesn't matter which comes first the "control"
samplesheet$condition <- as.factor(samplesheet$sample)

samplesheet$condition <- 
?mutate


# Now that the samplesheet is ready to used in DESeq2, 
# we'll write it out for future use.
write_rds(samplesheet, "final_samplesheet.rds")
```


(2) run deseq to create a dds (DESeqDataSet) 
this requires a matrix of rounded counts (made above)
::  DESeq(dds)

(3) normalize the counts to "rlog_counts" ::
rlog(dds, blind = TRUE)

# Important info on this here:
https://compbiocore.github.io/deseq-workshop-1/assets/deseq_workshop_1.html

(4) retreive rlog counts using :: assay(rlog_counts)

```{R essential set up to running DeSeq}

# (1) first run DESeq2 by creating a dds object.
# We use the function "DESeqDataSetFromMatrix"
# With parameters of countData, colData, design

dds <- DESeqDataSetFromMatrix(countData = counts_filtered,
                              # this is our counts data we read in
                              colData = samplesheet,
                              # telling DeSeq what env variable to use for sample sheet
                              design = ~ condition)
                              # perhaps most important is "condition" is a factor in samplesheet 

# (2) run DESeq2 function on dds object
dds <- DESeq(dds)

# (3) Normalize counts (rlog)
# This basically is rank counts normalized to std error in replicates.
rlog_counts <- rlog(dds, blind = TRUE)

# (4) now we retrieve the values using the "assay" function that converts to rlog_counts)
rlog_counts_matrix <- assay(rlog_counts)

# We now have a bunch more results embedded in the dds object
resultsNames(dds)



# Now we can write this out and START from here in the future.
# create new folder 'rlog_count_matrices'
write_rds(rlog_counts_matrix, "results/rlog_counts_all.rds")

```

# Let's look at the results !
```{r}

res <- results(dds, name = "X")

# We can further index the information for just this sample:
res_df <- res %>% as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  merge(g2s) %>%
  mutate(result_name = "condition_membrane_fraction_vs_total")

# Looking to see log2 fold change range
summary(res_df$log2FoldChange)
hist(res_df$log2FoldChange)

```

# Above we did all-vs-all conditions 
# We could also factor the sample sheet to have all timepoints
# versus 0

#TODO: Mingfeng

```{r}

```


# Now adding Linear model to test across time
# Linear model doesn't make sense since all we have is dox treated over time
```{r}

wt_overexp_long_dds <- DESeqDataSetFromMatrix(countData = counts_filtered, 
                                          colData = wt_overexp_long_samples, 
                                          design = ~ firre_induced + timepoint + timepoint*firre_induced)

```